### PR TITLE
Increment penalty in case of repeated violation during previous duration

### LINKED ISSUE #259: penaltyPercent to scale with each violation
Add logic to TACoApplication.sol such that the `penaltyPercent` parameter increases with every violation committed by a given staker. 

To make it as simple as possible in the first version:  
- All incomplete DKG rituals are treated as equal violations regardless of the reason or context. In future versions there could be varying punishment/severity for: 
   - Incomplete DKGs due to being offline vs. not running an up-to-date version vs. other known failures 
   - Incomplete DKGs where more than one node is affected and/or there is correlation (e.g. cloud provider goes down). Note this could lead to more severe punishments to encourage diverse infrastructure. 
   - Incomplete DKGs that indicate malicious behavior 
   - Incomplete handover DKGs during deauth periods. 
- `penaltyDuration` should be a constant 
   - e.g. 1 month of reduced rewards for all violations
- A violation that occurs during an already penalized reward period simply restarts the 1 month clock from the day the violation is logged (but with increased severity). In other words, committing a violation before a previous punishment has ended does not lead to any special outcomes, like summing `penaltyPercent`s. 
- A staker's violation count resets after a universal fixed duration (e.g. 24 months) from their first violation, after which point a fresh violation would result in the least severe `penaltyPercent`. 

Hypothetical initial parameters: 
- First incomplete DKG violation leads to a 25% reduction in rewards for a month. 
- The second leads to a 50%
- Third to a 75%, 
- Fourth (& fifth, sixth...nth) to a 100% reduction. 

Depending on resultant behavior, we can consider increasing `penaltyDuration` for violations beyond the fourth. The rationale for avoiding this in the first version is that we don't want extended periods where nodes are earning nothing, as that may encourage them to exit the network. 

Comment from @manumonti: 
About the penalty duration increasing beyond the fourth: I think we should consider setting its max value to the deauthorization delay (183 days) or less. If this reset duration is much higher than the deauth delay, the [reoffending] operators with 100% of the penalty will be incentivized to just remove the operator from the network and set up a new one with a different staking provider address to avoid the penalty. And this can lead to an unstable network with cohorts with unreachable nodes.

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
See #259

**Issues fixed/closed:**
Fixes #259 

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

### CODE REVIEW

**Review by manumonti — APPROVED** [review]
LGTM! 👌

**Review by derekpierre — APPROVED** [review]
🎸

File: contracts/contracts/TACoApplication.sol [hunk 1]
```diff
@@ -244,6 +245,7 @@ contract TACoApplication is
      * @param _commitmentDeadline Last date to make a commitment
      * @param _penaltyDefault Default penalty percentage (as a value out of 10000)
      * @param _penaltyDuration Duration of penalty
+     * @param _penaltyIncrement Increment of penalty if it happens during previous penalty duration
```
Comment (arjunhassard) [inline contracts/contracts/TACoApplication.sol:248]:
```suggestion
     * @param _penaltyIncrement Increment of penalty if violation occurs during an existing penalty period
```
Just to clarify that 'previous' penalties have no bearing on this variable

**Review by arjunhassard — APPROVED** [review]
Nice work! One future-proofing comment; at some point it would be useful to add a variable – `violationTimeframe` or something, which can be tuned so that if there are lots of DKG incompletions within a short timeframe (e.g. 3 hours), the guilty staker is given a reprieve from penalty increments, and only sees their rewards reduced by the default (unless they were already within a penalty period, of course!).

**Review by cygnusv — APPROVED** [review]

**Review by cygnusv — APPROVED** [review]

**Pull request merged**